radio/src/gui/colorlcd/themes/etx_lv_theme.cpp

```diff
@@ -11,1195 +27,769 @@
+    .height_def = LV_DPI_DEF,
+    .editable = LV_OBJ_CLASS_EDITABLE_FALSE,
+    .group_def = LV_OBJ_CLASS_GROUP_DEF_FALSE,
+    .instance_size = sizeof(lv_obj_t)
+};
+
+const lv_obj_class_t table_class  = {
+    .base_class = &lv_table_class,
+    .constructor_cb = nullptr,
+    .destructor_cb = nullptr,
+    .user_data = nullptr,
+    .event_cb = table_event,
+    .width_def = LV_SIZE_CONTENT,
+    .height_def = LV_SIZE_CONTENT,
+    .editable = LV_OBJ_CLASS_EDITABLE_TRUE,
+    .group_def = LV_OBJ_CLASS_GROUP_DEF_TRUE,
+    .instance_size = sizeof(lv_table_t),
+};
+
+const lv_obj_class_t etx_switch_class = {
+    .base_class = &lv_switch_class,
+    .constructor_cb = nullptr,
+    .destructor_cb = nullptr,
+    .user_data = nullptr,
+    .event_cb = nullptr,
+    .width_def = 0,
+    .height_def = 32,
+    .editable = LV_OBJ_CLASS_EDITABLE_INHERIT,
+    .group_def = LV_OBJ_CLASS_GROUP_DEF_INHERIT,
+    .instance_size = sizeof(lv_switch_t),
+};
+
+const lv_obj_class_t etx_button_class = {
+    .base_class = &lv_btn_class,
+    .constructor_cb = nullptr,
+    .destructor_cb = nullptr,
+    .user_data = nullptr,
+    .event_cb = nullptr,
+    .width_def = LV_SIZE_CONTENT,
+    .height_def = 32,
+    .editable = LV_OBJ_CLASS_EDITABLE_INHERIT,
+    .group_def = LV_OBJ_CLASS_GROUP_DEF_TRUE,
+    .instance_size = sizeof(lv_btn_t),
+};
+
+const lv_obj_class_t etx_choice_class = {
+    .base_class = &lv_obj_class,
+    .constructor_cb = nullptr,
+    .destructor_cb = nullptr,
+    .user_data = nullptr,
+    .event_cb = nullptr,
+    .width_def = LV_SIZE_CONTENT,
+    .height_def = 32,
+    .editable = LV_OBJ_CLASS_EDITABLE_INHERIT,
+    .group_def = LV_OBJ_CLASS_GROUP_DEF_INHERIT,
+    .instance_size = sizeof(lv_obj_t),
+};
+
+const lv_obj_class_t etx_bar_class = {
+    .base_class = &lv_bar_class,
+    .constructor_cb = nullptr,
+    .destructor_cb = nullptr,
+    .user_data = nullptr,
+    .event_cb = nullptr,
+    .width_def = lv_pct(100),
+    .height_def = 32,
+    .editable = LV_OBJ_CLASS_EDITABLE_INHERIT,
+    .group_def = LV_OBJ_CLASS_GROUP_DEF_INHERIT,
+    .instance_size = sizeof(lv_bar_t),
+};
+
+lv_obj_t* input_mix_line_create(lv_obj_t* parent)
 {
-    if(!etx_lv_theme_is_inited()) {
-        return NULL;
-    }
+  lv_obj_t * obj = lv_obj_class_create_obj(&input_mix_line_class, parent);
+  lv_obj_class_init_obj(obj);
+
+  lv_obj_add_style(obj, &styles.line_btn, 0);
+  lv_obj_add_style(obj, &styles.rounded, 0);
+  lv_obj_add_style(obj, &styles.pad_tiny, 0);
+  lv_obj_set_style_bg_color(obj, makeLvColor(COLOR_THEME_SECONDARY3), LV_STATE_CHECKED);
+  lv_obj_add_style(obj, &styles.focus_border, LV_STATE_FOCUSED);
+  lv_obj_set_style_text_color(obj, makeLvColor(COLOR_THEME_SECONDARY1), 0);
+  lv_obj_set_style_text_font(obj, getFont(FONT(STD)), 0);
+
+  return obj;
+}
+
+lv_obj_t* input_mix_group_create(lv_obj_t* parent)
+{
+  lv_obj_t * obj = lv_obj_class_create_obj(&input_mix_group_class, parent);
+  lv_obj_class_init_obj(obj);
 
```

Comment:
For all object that are created by those create functions a style should be created once, i.e when the theme is initialized or when another EdgeTX theme is selected and this should then be applied on object creation. Else LVGL will start to render the page with every ```lv_obj_add_style``` or ```lv_obj_set_style...``` instead of just once at object creation. It might even make sense to try to find a way to have the style applied once before the object is registered with LVGL, that way there should be only one page invalidation per object creation. Even with one style being applied avter ```lv_obj_class_create_obj``` or ```lv_obj_create``` the page is probably invalidated twice.

Reply:
Apart from a couple I missed, everything that can be done with styles should be. The styles are created once and only updated if the theme changes (and only the colors are updated).

Reply:
I've incorporated the object constructor pattern as suggested - thanks for that.